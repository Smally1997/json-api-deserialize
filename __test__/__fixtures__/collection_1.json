{
  "data": [
    {
      "id": "123",
      "type": "fundraiser",
      "attributes": {
        "updated_at": "2019-06-24T17:49:16.000Z",
        "created_at": "2019-05-09T15:28:44.000Z",
        "name": "Hogwarts Varsity Quidditch Team 2019"
      },
      "relationships": {
        "contact": {
          "data": {
            "id": "11650",
            "type": "contact"
          }
        },
        "organization": {
          "data": {
            "id": "19925",
            "type": "organization"
          }
        },
        "events": {
          "data": [
            {
              "id": "351978",
              "type": "event"
            },
            {
              "id": "351979",
              "type": "event"
            }
          ]
        }
      }
    },
    {
      "id": "1234",
      "type": "fundraiser",
      "attributes": {
        "updated_at": "2019-06-24T17:49:16.000Z",
        "created_at": "2019-05-09T15:28:44.000Z",
        "name": "Hogwarts JV Quidditch Team 2019"
      },
      "relationships": {
        "contact": {
          "data": {
            "id": "11651",
            "type": "contact"
          }
        },
        "organization": {
          "data": {
            "id": "19926",
            "type": "organization"
          }
        },
        "events": {
          "data": [
            {
              "id": "351980",
              "type": "event"
            },
            {
              "id": "351981",
              "type": "event"
            }
          ]
        }
      }
    }
  ],
  "included": [
    {
      "id": "27284",
      "type": "address",
      "attributes": {
        "updated_at": "2017-12-20T22:23:58.000Z",
        "created_at": "2017-12-20T22:23:58.000Z",
        "street": "123 Banffshire Ave",
        "city": "Dufftown",
        "region": "Scottish Highlands",
        "country": "Scotland",
        "zip": "12345",
        "abbreviated_region": "wa"
      }
    },
    {
      "id": "19925",
      "type": "organization",
      "attributes": {
        "updated_at": "2019-06-24T18:35:19.000Z",
        "created_at": "2018-11-20T21:41:49.000Z",
        "name": "Hogwarts Academy"
      },
      "relationships": {
        "address": {
          "data": {
            "id": "27284",
            "type": "address"
          }
        }
      }
    },
    {
      "id": "351978",
      "type": "event",
      "attributes": {
        "updated_at": "2019-06-21T18:30:25.000Z",
        "created_at": "2019-06-05T02:06:54.000Z",
        "name": "Training with Prof Snape",
        "status": "active"
      }
    },
    {
      "id": "351979",
      "type": "event",
      "attributes": {
        "updated_at": "2019-06-21T18:30:25.000Z",
        "created_at": "2019-06-05T02:06:54.000Z",
        "name": "Training with Prof Snape 2",
        "status": "pending"
      }
    },
    {
      "id": "27285",
      "type": "address",
      "attributes": {
        "updated_at": "2017-12-20T22:23:58.000Z",
        "created_at": "2017-12-20T22:23:58.000Z",
        "street": "124 Banffshire Ave",
        "city": "Dufftown 2",
        "region": "Scottish Highlands 2",
        "country": "Scotland 2",
        "zip": "12346",
        "abbreviated_region": "oh"
      }
    },
    {
      "id": "19926",
      "type": "organization",
      "attributes": {
        "updated_at": "2019-06-24T18:35:19.000Z",
        "created_at": "2018-11-20T21:41:49.000Z",
        "name": "Hogwarts Academy 2"
      },
      "relationships": {
        "address": {
          "data": {
            "id": "27285",
            "type": "address"
          }
        }
      }
    },
    {
      "id": "351980",
      "type": "event",
      "attributes": {
        "updated_at": "2019-06-21T18:30:25.000Z",
        "created_at": "2019-06-05T02:06:54.000Z",
        "name": "Training with Prof Snape 2",
        "status": "active"
      }
    },
    {
      "id": "351981",
      "type": "event",
      "attributes": {
        "updated_at": "2019-06-21T18:30:25.000Z",
        "created_at": "2019-06-05T02:06:54.000Z",
        "name": "Training with Prof Snape 3",
        "status": "pending"
      }
    }
  ]
}