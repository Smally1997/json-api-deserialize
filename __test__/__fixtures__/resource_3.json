{
  "data": {
    "type": "building",
    "id": "1",
    "attributes": {
      "name": "Mercer Island Books"
    },
    "relationships": {
      "owner": {
        "data": {
          "id": "2",
          "type": "owner"
        }
      }
    }
  },
  "included": [
    {
      "type": "owner",
      "id": "2",
      "attributes": {
        "name": "John Stamos"
      },
      "relationships": {
        "address": {
          "data": {
            "id": "3",
            "type": "address"
          }
        }
      }
    },
    {
      "type": "address",
      "id": "3",
      "attributes": {
        "street": "123 San Fran Baby Ave",
        "type": "commercial",
        "zip_code": "12345"
      },
      "relationships": {
        "driveway": {
          "data": {
            "id": "4",
            "type": "driveway"
          }
        }
      }
    },
    {
      "type": "driveway",
      "id": "4",
      "attributes": {
        "type": "The fancy circle type with a fountain in the middle"
      }
    }
  ]
}
