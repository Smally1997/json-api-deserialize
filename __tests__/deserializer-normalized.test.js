const fs = require('fs');
const path = require('path');

const deserialize = require('../dist').deserialize;

// Fixtures
const resourceResponses = [
  require('./__fixtures__/resource_1.json'),
  require('./__fixtures__/resource_2.json'),
  require('./__fixtures__/resource_3.json'),
  require('./__fixtures__/resource_4.json'),
  require('./__fixtures__/resource_5.json'),
  require('./__fixtures__/resource_6.json'),
  require('./__fixtures__/resource_7.json'),
  require('./__fixtures__/resource_8.json'),
];

const collectionResponses = [
  require('./__fixtures__/collection_1.json'),
  require('./__fixtures__/collection_2.json'),
];

describe('JsonApiDeserializer', () => {
  resourceResponses.forEach((resourceResponse, index) => {
    describe(`Resource Response (${index + 1})`, () => {
      const normalizedResponse = deserialize(resourceResponse, { normalize: true });

      fs.writeFileSync(
        path.join(__dirname, `./__outputs__/resource_normalized_${index + 1}.json`),
        JSON.stringify(normalizedResponse)
      );

      test('RO data object contains nested full resource objects', () => {
        expect(normalizedResponse).toMatchSnapshot();
      });
    });
  });

  collectionResponses.forEach((collectionResponse, index) => {
    describe(`Collection Response (${index + 1})`, () => {
      const normalizedResponse = deserialize(collectionResponse, { normalize: true });

      fs.writeFileSync(
        path.join(__dirname, `./__outputs__/collection_normalized_${index + 1}.json`),
        JSON.stringify(normalizedResponse)
      );

      test('RO data object contains nested full resource objects', () => {
        expect(normalizedResponse).toMatchSnapshot();
      });
    });
  });
});
