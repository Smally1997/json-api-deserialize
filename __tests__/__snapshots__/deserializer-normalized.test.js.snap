// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsonApiDeserializer Collection Response (1) RO data object contains nested full resource objects 1`] = `
Object {
  "data": Array [
    Object {
      "contact": Object {
        "id": "11650",
      },
      "createdAt": "2019-05-09T15:28:44.000Z",
      "events": Array [
        Object {
          "createdAt": "2019-06-05T02:06:54.000Z",
          "id": "351978",
          "name": "Training with Prof Snape",
          "status": "active",
          "updatedAt": "2019-06-21T18:30:25.000Z",
        },
        Object {
          "createdAt": "2019-06-05T02:06:54.000Z",
          "id": "351979",
          "name": "Training with Prof Snape 2",
          "status": "pending",
          "updatedAt": "2019-06-21T18:30:25.000Z",
        },
      ],
      "id": "123",
      "name": "Hogwarts Varsity Quidditch Team 2019",
      "organization": Object {
        "address": Object {
          "abbreviatedRegion": "wa",
          "city": "Dufftown",
          "country": "Scotland",
          "createdAt": "2017-12-20T22:23:58.000Z",
          "id": "27284",
          "region": "Scottish Highlands",
          "street": "123 Banffshire Ave",
          "updatedAt": "2017-12-20T22:23:58.000Z",
          "zip": "12345",
        },
        "createdAt": "2018-11-20T21:41:49.000Z",
        "id": "19925",
        "name": "Hogwarts Academy",
        "updatedAt": "2019-06-24T18:35:19.000Z",
      },
      "updatedAt": "2019-06-24T17:49:16.000Z",
    },
    Object {
      "contact": Object {
        "id": "11651",
      },
      "createdAt": "2019-05-09T15:28:44.000Z",
      "events": Array [
        Object {
          "createdAt": "2019-06-05T02:06:54.000Z",
          "id": "351980",
          "name": "Training with Prof Snape 2",
          "status": "active",
          "updatedAt": "2019-06-21T18:30:25.000Z",
        },
        Object {
          "createdAt": "2019-06-05T02:06:54.000Z",
          "id": "351981",
          "name": "Training with Prof Snape 3",
          "status": "pending",
          "updatedAt": "2019-06-21T18:30:25.000Z",
        },
      ],
      "id": "1234",
      "name": "Hogwarts JV Quidditch Team 2019",
      "organization": Object {
        "address": Object {
          "abbreviatedRegion": "oh",
          "city": "Dufftown 2",
          "country": "Scotland 2",
          "createdAt": "2017-12-20T22:23:58.000Z",
          "id": "27285",
          "region": "Scottish Highlands 2",
          "street": "124 Banffshire Ave",
          "updatedAt": "2017-12-20T22:23:58.000Z",
          "zip": "12346",
        },
        "createdAt": "2018-11-20T21:41:49.000Z",
        "id": "19926",
        "name": "Hogwarts Academy 2",
        "updatedAt": "2019-06-24T18:35:19.000Z",
      },
      "updatedAt": "2019-06-24T17:49:16.000Z",
    },
  ],
  "deserialized": true,
}
`;

exports[`JsonApiDeserializer Collection Response (2) RO data object contains nested full resource objects 1`] = `
Object {
  "data": Array [
    Object {
      "author": Object {
        "firstName": "Dan",
        "id": "9",
        "lastName": "Gebhardt",
        "links": Object {
          "self": "http://example.com/people/9",
        },
        "twitter": "dgeb",
        "type": "human",
      },
      "comments": Array [
        Object {
          "author": Object {
            "id": "2",
          },
          "body": "First!",
          "id": "5",
          "links": Object {
            "self": "http://example.com/comments/5",
          },
        },
        Object {
          "author": Object {
            "firstName": "Dan",
            "id": "9",
            "lastName": "Gebhardt",
            "links": Object {
              "self": "http://example.com/people/9",
            },
            "twitter": "dgeb",
            "type": "human",
          },
          "body": "I like XML better",
          "id": "12",
          "links": Object {
            "self": "http://example.com/comments/12",
          },
        },
      ],
      "id": "1",
      "links": Object {
        "self": "http://example.com/articles/1",
      },
      "title": "JSON:API paints my bikeshed!",
    },
  ],
  "deserialized": true,
}
`;

exports[`JsonApiDeserializer Resource Response (1) RO data object contains nested full resource objects 1`] = `
Object {
  "data": Object {
    "contact": Object {
      "id": "11650",
    },
    "createdAt": "2019-05-09T15:28:44.000Z",
    "events": Array [
      Object {
        "createdAt": "2019-06-05T02:06:54.000Z",
        "id": "351978",
        "name": "Training with Prof Snape",
        "status": "active",
        "updatedAt": "2019-06-21T18:30:25.000Z",
      },
      Object {
        "createdAt": "2019-06-05T02:06:54.000Z",
        "id": "351979",
        "name": "Training with Prof Snape 2",
        "status": "pending",
        "updatedAt": "2019-06-21T18:30:25.000Z",
      },
    ],
    "id": "123",
    "name": "Hogwarts Varsity Quidditch Team 2019",
    "organization": Object {
      "address": Object {
        "abbreviatedRegion": "wa",
        "city": "Dufftown",
        "country": "Scotland",
        "createdAt": "2017-12-20T22:23:58.000Z",
        "id": "27284",
        "region": "Scottish Highlands",
        "street": "123 Banffshire Ave",
        "updatedAt": "2017-12-20T22:23:58.000Z",
        "zip": "12345",
      },
      "createdAt": "2018-11-20T21:41:49.000Z",
      "id": "19925",
      "name": "Hogwarts Academy",
      "updatedAt": "2019-06-24T18:35:19.000Z",
    },
    "updatedAt": "2019-06-24T17:49:16.000Z",
  },
  "deserialized": true,
}
`;

exports[`JsonApiDeserializer Resource Response (2) RO data object contains nested full resource objects 1`] = `
Object {
  "data": Object {
    "author": Object {
      "age": 80,
      "gender": "male",
      "id": "42",
      "name": "John",
    },
    "body": "The shortest article. Ever.",
    "createdAt": "2015-05-22T14:56:29.000Z",
    "id": "1",
    "title": "JSON:API paints my bikeshed!",
    "updatedAt": "2015-05-22T14:56:28.000Z",
  },
  "deserialized": true,
}
`;

exports[`JsonApiDeserializer Resource Response (3) RO data object contains nested full resource objects 1`] = `
Object {
  "data": Object {
    "createdAt": "2019-06-30T22:33:41+0000",
    "id": "1",
    "name": "Mercer Island Book Store",
    "owner": Object {
      "address": Object {
        "createdAt": "2019-06-30T22:33:41+0000",
        "driveway": Object {
          "createdAt": "2019-06-30T22:33:41+0000",
          "id": "4",
          "type": "parking_lot",
          "updatedAt": "2019-06-30T22:33:41+0000",
        },
        "id": "3",
        "street": "123 Test Ave",
        "type": "commercial",
        "updatedAt": "2019-06-30T22:33:41+0000",
        "zipCode": "12345",
      },
      "createdAt": "2019-06-30T22:33:41+0000",
      "fullName": "John Doe",
      "id": "2",
      "updatedAt": "2019-06-30T22:33:41+0000",
    },
    "updatedAt": "2019-06-30T22:33:41+0000",
  },
  "deserialized": true,
}
`;

exports[`JsonApiDeserializer Resource Response (4) RO data object contains nested full resource objects 1`] = `
Object {
  "data": Object {
    "author": Object {
      "id": "42",
    },
    "body": "The shortest article. Ever.",
    "createdAt": "2015-05-22T14:56:29.000Z",
    "id": "1",
    "title": "JSON:API paints my bikeshed!",
    "updatedAt": "2015-05-22T14:56:28.000Z",
  },
  "deserialized": true,
  "meta": Object {
    "foo": "bar",
  },
}
`;

exports[`JsonApiDeserializer Resource Response (5) RO data object contains nested full resource objects 1`] = `
Object {
  "data": Object {
    "author": Object {
      "links": Object {
        "related": "http://example.com/articles/1/author",
      },
    },
    "id": "1",
    "title": "JSON:API paints my bikeshed!",
  },
  "deserialized": true,
  "links": Object {
    "self": "http://example.com/articles/1",
  },
}
`;

exports[`JsonApiDeserializer Resource Response (6) RO data object contains nested full resource objects 1`] = `
Object {
  "data": null,
  "deserialized": true,
  "links": Object {
    "self": "http://example.com/articles/1/author",
  },
}
`;

exports[`JsonApiDeserializer Resource Response (7) RO data object contains nested full resource objects 1`] = `
Object {
  "data": Array [],
  "deserialized": true,
  "links": Object {
    "related": "/articles/1/tags",
    "self": "/articles/1/relationships/tags",
  },
}
`;

exports[`JsonApiDeserializer Resource Response (8) RO data object contains nested full resource objects 1`] = `
Object {
  "data": Object {
    "id": "550e8400-e29b-41d4-a716-446655440000",
    "links": Object {
      "self": "http://example.com/photos/550e8400-e29b-41d4-a716-446655440000",
    },
    "src": "http://example.com/images/productivity.png",
    "title": "Ember Hamster",
  },
  "deserialized": true,
}
`;
